src/Mvc/test/Microsoft.AspNetCore.Mvc.FunctionalTests/ComponentRenderingFunctionalTests.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using AngleSharp.Parser.Html;
+using BasicWebSite.Services;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.FunctionalTests
+{
+    public class ComponentRenderingFunctionalTests : IClassFixture<MvcTestFixture<BasicWebSite.Startup>>
+    {
+        public ComponentRenderingFunctionalTests(MvcTestFixture<BasicWebSite.Startup> fixture)
+        {
+            Client = fixture.WithWebHostBuilder(b => b.ConfigureServices(ConfigureServices)).CreateClient();
+            LoopHandler.InnerHandler = fixture.Server.CreateHandler();
+
+            void ConfigureServices(IServiceCollection services) =>
+                services.AddSingleton(new WeatherForecastService(new HttpClient(LoopHandler)
+                {
+                    BaseAddress = fixture.ClientOptions.BaseAddress
+                }));
+        }
+
+        private class LoopHttpHandler : DelegatingHandler
+        {
+        }
+
+        public HttpClient Client { get; }
+
+        private LoopHttpHandler LoopHandler { get; set; } = new LoopHttpHandler();
+
+        [Fact]
+        public async Task Renders_BasicComponent()
+        {
+            // Arrange & Act
+            var response = await Client.GetAsync("http://localhost/components");
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var content = await response.Content.ReadAsStringAsync();
+
+            AssertComponent("\n    <p>Hello world!</p>\n", "Greetings", content);
+        }
+
+        [Fact]
+        public async Task Renders_AsyncComponent()
+        {
+            // Arrange & Act
+            var expectedHtml = string.Join(
```

Comment:
Can this instead be,
```c#
var expectedHtml = @"
<h1>Weather forecast</h1>"
.
.
.
";
```

It gets very hard to maintain with string.join and a bunch of newlines embedded in the string

Reply:
I can't. The compiler produces specific line endings wich are different from the ones a literal string will produce.